File: 2780-make-costs-of-paths-equal-in-a-binary-tree/solution.ts
function minIncrements(n: number, cost: number[]): number {
  let answer = 0;

  while (n > 1) {
    const start = (n + 1) / 2 - 1;
    for (let i = start; i < n; i += 2) {
      const parent = Math.floor((i - 1) / 2);
      answer += Math.max(cost[i], cost[i + 1]) - Math.min(cost[i], cost[i + 1]);
      cost[parent] += Math.max(cost[i], cost[i + 1]);
    }
    n = (n + 1) / 2 - 1;
  }

  return answer;
}

File: 2780-make-costs-of-paths-equal-in-a-binary-tree/solution.cpp
class Solution {
public:
    int minIncrements(int n, vector<int>& cost) {
        int ans = 0;
        vector<int> child(n, 0);
        for(int start = (n + 1) / 2; start >= 2; start /= 2) {
            for(int i = start; i < start * 2; i += 2) {
                int ci = cost[i - 1] + child[i - 1], cj = cost[i] + child[i];
                ans += abs(ci - cj);
                child[i / 2 - 1] = max(ci, cj);
            }
        }
        return ans;
    }
};
